modular_skyrat/modules/imenu/code/imenu.dm

```diff
@@ -0,0 +1,268 @@
+
+/proc/populate_interaction_jsons_master(path)
+	if(!fexists(path))
+		message_admins("We are attempting to load an interaction master without the file existing! '[path]'")
+		return
+	var/file = file(path)
+	var/list/json = json_load(file)
+
+	for(var/iname in json)
+		if(GLOB.interaction_instances[iname])
+			message_admins("Interaction Master '[path]' contained a duplicate interaction! '[iname]'")
+			continue
+
+		var/list/ijson = json[iname]
+		if(ijson["name"] != iname)
+			message_admins("Interaction Master '[path]' contained an invalid interaction! '[iname]'")
+			continue
+
+		var/datum/interaction/int = new()
+
+		int.distance_allowed = sanitize_integer(ijson["distance_allowed"], 0, 1, 0)
+		int.message = sanitize_islist(ijson["message"], list("json error"))
+		int.category = sanitize_text(ijson["category"])
+		int.usage = sanitize_text(ijson["usage"])
+		int.sound_use = sanitize_integer(ijson["sound_use"], 0, 1, 0)
+		int.sound_range = sanitize_integer(ijson["sound_range"], 1, 7, 1)
+		int.sound_possible = sanitize_islist(ijson["sound_possible"], list("json error"))
+		int.interaction_requires = sanitize_islist(ijson["interaction_requires"], list())
+
+		GLOB.interaction_instances[iname] = int
+
+/datum/interaction/proc/load_from_json(path)
+	var/fpath = path
+	if(!fexists(fpath))
+		message_admins("Attempted to load an interaction from json and the file does not exist")
+		qdel(src)
+		return FALSE
+	var/file = file(fpath)
+	var/list/json = json_load(file)
+	name = sanitize_text(json["name"])
+	distance_allowed = sanitize_integer(json["distance_allowed"], 0, 1, 0)
+	message = sanitize_islist(json["message"], list("json error"))
+	category = sanitize_text(json["category"])
+	usage = sanitize_text(json["usage"])
+	sound_use = sanitize_integer(json["sound_use"], 0, 1, 0)
+	sound_range = sanitize_integer(json["sound_range"], 1, 7, 1)
+	sound_possible = sanitize_islist(json["sound_possible"], list("json error"))
+	interaction_requires = sanitize_islist(json["interaction_requires"], list())
+	return TRUE
+
+/datum/interaction/proc/json_save(path)
+	var/fpath = path
+	if(fexists(fpath))
+		fdel(fpath)
+	var/list/json = list(
+		"name" = name,
+		"distance_allowed" = distance_allowed,
+		"message" = message,
+		"category" = category,
+		"usage" = usage,
+		"sound_use" = sound_use,
+		"sound_range" = sound_range,
+		"sound_possible" = sound_possible,
+		"interaction_requires" = interaction_requires
+	)
+	var/file = file(fpath)
+	WRITE_FILE(file, json_encode(json))
+	return TRUE
+
+/datum/interaction
+	var/name = "broken interaction"
+	var/distance_allowed = FALSE
+	var/list/message = list()
+	var/category = INTERACTION_CAT_HIDE
+	var/usage = INTERACTION_OTHER
+	var/sound_use = FALSE
+	var/sound_range = 1
+	var/sound_cache = null
+	var/list/sound_possible = list()
+	var/list/interaction_requires = list()
+
+/datum/interaction/proc/allow_act(mob/living/user, mob/living/target)
```

Comment:
I feel like this could be done much better

Reply:
Even just skipping the switch and just checking for "does it have this flag"

Reply:
Bleh, the for

Reply:
I have the for to check all the requirements procedurally otherwise if one isn't implemented there will be no way to know